### ISSUE TITLE
BAD ACCESS Crashes on remote commands

### ISSUE THREAD

**Issue by Enricoza** [Tealium/tealium-swift#121]
**Your Environment**
Tealium Swift library version [e.g. 1.6.7]: 1.8.0-beta3

**Describe the bug**
I found Several crashes, all connected to the remote commands. I think it's all related to some threading issue, but once again I can't be sure since I didn't look up.

I'm adding the stack traces in order from the most impactful crash to the lowest.

```
#0
Crashed: com.tealium.backgroundserialqueue
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000
0  TealiumSwift                   0x1032a5ed8 specialized Dictionary._Variant.removeValue(forKey:) (<compiler-generated>)
1  TealiumSwift                   0x1032a626c specialized static TealiumRemoteCommand.sendCompletionNotification(for:response:) (<compiler-generated>)
2  TealiumSwift                   0x1032a89a4 closure #1 in TealiumRemoteCommands.tealiumRemoteCommandRequestsExecution(_:response:) + 145 (TealiumRemoteCommands.swift:145)
3  TealiumSwift                   0x103248ca8 thunk for @escaping @callee_guaranteed () -> () (<compiler-generated>)
4  libdispatch.dylib              0x1b1a4e610 _dispatch_call_block_and_release + 24
5  libdispatch.dylib              0x1b1a4f184 _dispatch_client_callout + 16
6  libdispatch.dylib              0x1b1a2c710 _dispatch_lane_serial_drain$VARIANT$armv81 + 564
7  libdispatch.dylib              0x1b1a2d128 _dispatch_lane_invoke$VARIANT$armv81 + 400
8  libdispatch.dylib              0x1b1a3643c _dispatch_workloop_worker_thread + 576
9  libsystem_pthread.dylib        0x1b1a9ef88 _pthread_wqthread + 276
10 libsystem_pthread.dylib        0x1b1aa1ad4 start_wqthread + 8
```

```
Crashed: com.tealium.backgroundserialqueue
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000
0  TealiumSwift                   0x106f6dd04 specialized Dictionary.subscript.getter (<compiler-generated>)
1  TealiumSwift                   0x106f6e214 specialized static TealiumRemoteCommand.sendCompletionNotification(for:response:) (<compiler-generated>)
2  TealiumSwift                   0x106f709a4 closure #1 in TealiumRemoteCommands.tealiumRemoteCommandRequestsExecution(_:response:) + 145 (TealiumRemoteCommands.swift:145)
3  TealiumSwift                   0x106f10ca8 thunk for @escaping @callee_guaranteed () -> () (<compiler-generated>)
4  libdispatch.dylib              0x1b52bd610 _dispatch_call_block_and_release + 24
5  libdispatch.dylib              0x1b52be184 _dispatch_client_callout + 16
6  libdispatch.dylib              0x1b529b73c _dispatch_lane_serial_drain$VARIANT$armv81 + 564
7  libdispatch.dylib              0x1b529c154 _dispatch_lane_invoke$VARIANT$armv81 + 400
8  libdispatch.dylib              0x1b52a543c _dispatch_workloop_worker_thread + 576
9  libsystem_pthread.dylib        0x1b530dfa4 _pthread_wqthread + 276
10 libsystem_pthread.dylib        0x1b5310ae0 start_wqthread + 8
```

```
Crashed: com.apple.main-thread
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000
0  libswiftCore.dylib             0x1b195911c swift_isUniquelyReferenced_nonNull_native + 34
1  TealiumSwift                   0x102d7c5dc TealiumRemoteCommands.triggerCommandFrom(request:) (<compiler-generated>)
2  TealiumSwift                   0x102d7c26c TealiumRemoteCommands.triggerCommandFrom(notification:) + 89 (TealiumRemoteCommands.swift:89)
3  TealiumSwift                   0x102d8093c partial apply for closure #1 in TealiumRemoteCommandsModule.enableNotifications() (<compiler-generated>)
4  TealiumSwift                   0x102d7f24c thunk for @escaping @callee_guaranteed (@in_guaranteed Notification) -> () (<compiler-generated>)
5  Foundation                     0x1a482fe74 -[__NSObserver _doit:] + 316
6  CoreFoundation                 0x1a44ac09c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 20
7  CoreFoundation                 0x1a44ac0e4 ___CFXRegistrationPost1_block_invoke + 64
8  CoreFoundation                 0x1a44ab444 _CFXRegistrationPost1 + 368
9  CoreFoundation                 0x1a44ab0fc ___CFXNotificationPost_block_invoke + 104
10 CoreFoundation                 0x1a442743c -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1416
11 CoreFoundation                 0x1a44aaa4c _CFXNotificationPost + 1244
12 TealiumSwift                   0x102d99308 specialized TealiumTagManagementWKWebView.webView(_:decidePolicyFor:decisionHandler:) (<compiler-generated>)
13 TealiumSwift                   0x102d973d8 @objc TealiumTagManagementWKWebView.webView(_:decidePolicyFor:decisionHandler:) (<compiler-generated>)
14 WebKit                         0x1abbef638 WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction(WebKit::WebPageProxy&, WTF::Ref<API::NavigationAction, WTF::DumbPtrTraits<API::NavigationAction> >&&, WTF::Ref<WebKit::WebFramePolicyListenerProxy, WTF::DumbPtrTraits<WebKit::WebFramePolicyListenerProxy> >&&, API::Object*) + 1004
15 WebKit                         0x1abc3c888 WebKit::WebPageProxy::decidePolicyForNavigationAction(WTF::Ref<WebKit::WebProcessProxy, WTF::DumbPtrTraits<WebKit::WebProcessProxy> >&&, WebKit::WebFrameProxy&, WebCore::SecurityOriginData&&, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, WTF::Ref<WebKit::WebPageProxy::PolicyDecisionSender, WTF::DumbPtrTraits<WebKit::WebPageProxy::PolicyDecisionSender> >&&) + 2384
16 WebKit                         0x1abc3beb8 WebKit::WebPageProxy::decidePolicyForNavigationActionAsyncShared(WTF::Ref<WebKit::WebProcessProxy, WTF::DumbPtrTraits<WebKit::WebProcessProxy> >&&, unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long) + 352
17 WebKit                         0x1abc3bd20 WebKit::WebPageProxy::decidePolicyForNavigationActionAsync(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long) + 144
18 WebKit                         0x1abe8d08c void IPC::callMemberFunctionImpl<WebKit::WebPageProxy, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long), std::__1::tuple<unsigned long long, WebCore::SecurityOriginData, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData, WebKit::FrameInfoData, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest, WebCore::ResourceRequest, IPC::FormDataReference, WebCore::ResourceResponse, WebKit::UserData, unsigned long long>, 0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul, 8ul, 9ul, 10ul, 11ul, 12ul>(WebKit::WebPageProxy*, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long), std::__1::tuple<unsigned long long, WebCore::SecurityOriginData, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData, WebKit::FrameInfoData, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest, WebCore::ResourceRequest, IPC::FormDataReference, WebCore::ResourceResponse, WebKit::UserData, unsigned long long>&&, std::__1::integer_sequence<unsigned long, 0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul, 8ul, 9ul, 10ul, 11ul, 12ul>) + 144
19 WebKit                         0x1abe8719c void IPC::handleMessage<Messages::WebPageProxy::DecidePolicyForNavigationActionAsync, WebKit::WebPageProxy, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long)>(IPC::Decoder&, WebKit::WebPageProxy*, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, WTF::Optional<WTF::ObjectIdentifier<WebCore::PageIdentifierType> >, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long)) + 84
20 WebKit                         0x1aba250e8 IPC::MessageReceiverMap::dispatchMessage(IPC::Connection&, IPC::Decoder&) + 104
21 WebKit                         0x1abc72630 WebKit::WebProcessProxy::didReceiveMessage(IPC::Connection&, IPC::Decoder&) + 32
22 WebKit                         0x1aba0f740 IPC::Connection::dispatchMessage(std::__1::unique_ptr<IPC::Decoder, std::__1::default_delete<IPC::Decoder> >) + 204
23 WebKit                         0x1aba12660 IPC::Connection::dispatchIncomingMessages() + 612
24 JavaScriptCore                 0x1b30d4a6c WTF::RunLoop::performWork() + 276
25 JavaScriptCore                 0x1b30d4d2c WTF::RunLoop::performWork(void*) + 36
26 CoreFoundation                 0x1a44ce7e0 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
27 CoreFoundation                 0x1a44ce738 __CFRunLoopDoSource0 + 80
28 CoreFoundation                 0x1a44cded0 __CFRunLoopDoSources0 + 180
29 CoreFoundation                 0x1a44c901c __CFRunLoopRun + 1080
30 CoreFoundation                 0x1a44c88bc CFRunLoopRunSpecific + 464
31 GraphicsServices               0x1ae334328 GSEventRunModal + 104
32 UIKitCore                      0x1a855e6d4 UIApplicationMain + 1936
33 Mediaset Play                  0x1000fe1b4 main + 24 (AppDelegate.swift:24)
34 libdyld.dylib                  0x1a4353460 start + 4
```

```
Crashed: com.tealium.backgroundserialqueue
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000
0  libswiftCore.dylib             0x1a28b111c swift_isUniquelyReferenced_nonNull_native + 34
1  TealiumSwift                   0x10592a26c specialized static TealiumRemoteCommand.sendCompletionNotification(for:response:) (<compiler-generated>)
2  TealiumSwift                   0x10592c9a4 closure #1 in TealiumRemoteCommands.tealiumRemoteCommandRequestsExecution(_:response:) + 145 (TealiumRemoteCommands.swift:145)
3  TealiumSwift                   0x1058ccca8 thunk for @escaping @callee_guaranteed () -> () (<compiler-generated>)
4  libdispatch.dylib              0x195175610 _dispatch_call_block_and_release + 24
5  libdispatch.dylib              0x195176184 _dispatch_client_callout + 16
6  libdispatch.dylib              0x19515373c _dispatch_lane_serial_drain$VARIANT$armv81 + 564
7  libdispatch.dylib              0x195154154 _dispatch_lane_invoke$VARIANT$armv81 + 400
8  libdispatch.dylib              0x19515d43c _dispatch_workloop_worker_thread + 576
9  libsystem_pthread.dylib        0x1951c5fa4 _pthread_wqthread + 276
10 libsystem_pthread.dylib        0x1951c8ae0 start_wqthread + 8
```

```
Crashed: com.apple.main-thread
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000
0  libswiftCore.dylib             0x1e87937ec swift_isUniquelyReferenced_nonNull_native + 10
1  TealiumSwift                   0x103685958 specialized Dictionary._Variant.setValue(_:forKey:) (<compiler-generated>)
2  TealiumSwift                   0x1036845dc TealiumRemoteCommands.triggerCommandFrom(request:) (<compiler-generated>)
3  TealiumSwift                   0x10368426c TealiumRemoteCommands.triggerCommandFrom(notification:) + 89 (TealiumRemoteCommands.swift:89)
4  TealiumSwift                   0x10368893c partial apply for closure #1 in TealiumRemoteCommandsModule.enableNotifications() (<compiler-generated>)
5  TealiumSwift                   0x10368724c thunk for @escaping @callee_guaranteed (@in_guaranteed Notification) -> () (<compiler-generated>)
6  Foundation                     0x1bba9bf20 <redacted> + 316
7  CoreFoundation                 0x1bb08a22c <redacted> + 20
8  CoreFoundation                 0x1bb08a1f8 <redacted> + 64
9  CoreFoundation                 0x1bb0896ec <redacted> + 392
10 CoreFoundation                 0x1bb089398 <redacted> + 96
11 CoreFoundation                 0x1bb002c64 <redacted> + 1496
12 CoreFoundation                 0x1bb088e48 _CFXNotificationPost + 696
13 TealiumSwift                   0x1036a1308 specialized TealiumTagManagementWKWebView.webView(_:decidePolicyFor:decisionHandler:) (<compiler-generated>)
14 TealiumSwift                   0x10369f3d8 @objc TealiumTagManagementWKWebView.webView(_:decidePolicyFor:decisionHandler:) (<compiler-generated>)
15 WebKit                         0x1ca77ce58 WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction(WebKit::WebPageProxy&, WTF::Ref<API::NavigationAction, WTF::DumbPtrTraits<API::NavigationAction> >&&, WTF::Ref<WebKit::WebFramePolicyListenerProxy, WTF::DumbPtrTraits<WebKit::WebFramePolicyListenerProxy> >&&, API::Object*) + 584
16 WebKit                         0x1ca7e3e0c WebKit::WebPageProxy::decidePolicyForNavigationAction(WTF::Ref<WebKit::WebProcessProxy, WTF::DumbPtrTraits<WebKit::WebProcessProxy> >&&, WebKit::WebFrameProxy&, WebCore::SecurityOriginData&&, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, unsigned long long, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, WTF::Ref<WebKit::WebPageProxy::PolicyDecisionSender, WTF::DumbPtrTraits<WebKit::WebPageProxy::PolicyDecisionSender> >&&) + 2572
17 WebKit                         0x1ca7e3398 WebKit::WebPageProxy::decidePolicyForNavigationActionAsyncShared(WTF::Ref<WebKit::WebProcessProxy, WTF::DumbPtrTraits<WebKit::WebProcessProxy> >&&, unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, unsigned long long, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long) + 376
18 WebKit                         0x1ca7e31e8 WebKit::WebPageProxy::decidePolicyForNavigationActionAsync(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, unsigned long long, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long) + 108
19 WebKit                         0x1ca9e85a0 void IPC::callMemberFunctionImpl<WebKit::WebPageProxy, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, unsigned long long, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long), std::__1::tuple<unsigned long long, WebCore::SecurityOriginData, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData, WebKit::FrameInfoData, unsigned long long, WebCore::ResourceRequest, WebCore::ResourceRequest, IPC::FormDataReference, WebCore::ResourceResponse, WebKit::UserData, unsigned long long>, 0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul, 8ul, 9ul, 10ul, 11ul, 12ul>(WebKit::WebPageProxy*, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, unsigned long long, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long), std::__1::tuple<unsigned long long, WebCore::SecurityOriginData, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData, WebKit::FrameInfoData, unsigned long long, WebCore::ResourceRequest, WebCore::ResourceRequest, IPC::FormDataReference, WebCore::ResourceResponse, WebKit::UserData, unsigned long long>&&, std::__1::integer_sequence<unsigned long, 0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul, 8ul, 9ul, 10ul, 11ul, 12ul>) + 108
20 WebKit                         0x1ca9e3078 void IPC::handleMessage<Messages::WebPageProxy::DecidePolicyForNavigationActionAsync, WebKit::WebPageProxy, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, unsigned long long, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long)>(IPC::Decoder&, WebKit::WebPageProxy*, void (WebKit::WebPageProxy::*)(unsigned long long, WebCore::SecurityOriginData&&, WebCore::PolicyCheckIdentifier, unsigned long long, WebKit::NavigationActionData&&, WebKit::FrameInfoData&&, unsigned long long, WebCore::ResourceRequest const&, WebCore::ResourceRequest&&, IPC::FormDataReference&&, WebCore::ResourceResponse&&, WebKit::UserData const&, unsigned long long)) + 100
21 WebKit                         0x1ca645e38 IPC::MessageReceiverMap::dispatchMessage(IPC::Connection&, IPC::Decoder&) + 120
22 WebKit                         0x1ca8058a8 WebKit::WebProcessProxy::didReceiveMessage(IPC::Connection&, IPC::Decoder&) + 32
23 WebKit                         0x1ca6369a8 IPC::Connection::dispatchMessage(std::__1::unique_ptr<IPC::Decoder, std::__1::default_delete<IPC::Decoder> >) + 148
24 WebKit                         0x1ca639498 IPC::Connection::dispatchIncomingMessages() + 712
25 JavaScriptCore                 0x1c22eb648 WTF::RunLoop::performWork() + 276
26 JavaScriptCore                 0x1c22eb908 WTF::RunLoop::performWork(void*) + 36
27 CoreFoundation                 0x1bb0aaf2c <redacted> + 24
28 CoreFoundation                 0x1bb0aaeac <redacted> + 88
29 CoreFoundation                 0x1bb0aa794 <redacted> + 176
30 CoreFoundation                 0x1bb0a56d0 <redacted> + 1004
31 CoreFoundation                 0x1bb0a4fc4 CFRunLoopRunSpecific + 436
32 GraphicsServices               0x1bd2a679c GSEventRunModal + 104
33 UIKitCore                      0x1e7383c38 UIApplicationMain + 212
34 Mediaset Play                  0x100b161b4 main + 24 (AppDelegate.swift:24)
35 libdyld.dylib                  0x1bab658e0 start + 4
```

**Expected behavior**
It shouldn't crash

**Additional context**
All those crashes are not very impactful singularly, but adding them together it's half the crashes on the outlined copy crash

**Comment by craigrouse** [comment]
Thanks for the details @Enricoza. Worth mentioning that if you're not actively using the Remote Commands module, then it is safe to disable/remove it, but I assume you are using it. We'll look into these right away and try to get you a quick resolution.

**Comment by Enricoza** [comment]
@craigrouse unfortunately I'm using the remote commands

**Comment by craigrouse** [comment]
@Enricoza as you have no doubt seen, 1.8.2 is now released, and should resolve this issue. We were able to reproduce this issue, but only with very high volumes of track requests, and those crashes went away after we implemented the fix, so hopefully this will be reflected in your app too when you update. We'll leave this open for now until you confirm that it's resolved. Thanks again for your valuable input.
